AX: Make all layout tests pass on macOS and iOS with ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE)

twilco opened the pull request:
#### 1e6dbc10af5dd7e55e331743bcec3e705b43480d
<pre>
AX: Make all layout tests pass on macOS and iOS with ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE)
<a href="https://bugs.webkit.org/show_bug.cgi?id=282647">https://bugs.webkit.org/show_bug.cgi?id=282647</a>
<a href="https://rdar.apple.com/139315447">rdar://139315447</a>

Reviewed by Chris Fleizach.

This patch features a variety of changes to make all layout tests pass with ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE):

  1. Prevent tables from returning &quot;invalid&quot; children — non-rows, columns, captions, or table header containers.
     We used to get this for free by way of AccessibilityTable::addChildren() only adding these types of children,
     but we cannot do that anymore, since the point of this effort is to add all children, including those that are
     ignored. Computing cell slots is now split out into a separate function: AccessibilityTable::computeCellSlots().

  2. When `inert` or `visibility` changes, re-compute AXPropertyName::IsIgnored for all descendants.

  3. Fix a bug where we don&apos;t auto-focus into modals if the author provided a focus element and immediately made
     that focus element un-focusable (e.g. via display:none). This was caught by a layout test, and was only exposed
     by this change by chance due to a slight timing difference.

  4. When an object becomes unignored, recompute all its properties, since we only cache a minimal set of properties for
     ignored objects.

  5. Split AccessibilityList::determineAccessibilityRole into determineAccessibilityRole and determineAccessibilityRoleWithCleanChildren
     The old implementation relied on children being up-to-date, without anything actually upholding that invariant. Now
     this is more explicit, and should behave more predictably.

  6. Fix a few ATSPI-platform guards added in <a href="https://bugs.webkit.org/show_bug.cgi?id=281749">https://bugs.webkit.org/show_bug.cgi?id=281749</a> that accidentally applied
     to iOS too.

  7. Drive-by fix to remove a completely unnecessary override for AccessibilityTableCell::parentObjectUnignored().

  8. Add new testing-only macOS attribute handling for `_AXIsInTable` in support of new test table-ancestry.html.

* LayoutTests/accessibility/aria-hidden-subtree-expected.txt:
* LayoutTests/accessibility/aria-hidden-subtree.html:
* LayoutTests/accessibility/aria-hidden-updates-alldescendants-expected.txt:
* LayoutTests/accessibility/aria-hidden-updates-alldescendants.html:
* LayoutTests/accessibility/aria-hidden-with-elements.html:
* LayoutTests/accessibility/aria-labelledby-on-password-input.html:
* LayoutTests/accessibility/aria-modal-in-aria-hidden-expected.txt:
* LayoutTests/accessibility/aria-modal-in-aria-hidden.html:
* LayoutTests/accessibility/deleting-iframe-destroys-axcache.html:
* LayoutTests/accessibility/dynamic-aria-hidden-cell-expected.txt:
* LayoutTests/accessibility/dynamic-aria-hidden-cell.html:
* LayoutTests/accessibility/ignore-modals-without-any-content.html:
* LayoutTests/accessibility/input-type-hidden-in-aria-hidden-false-expected.txt:
* LayoutTests/accessibility/input-type-hidden-in-aria-hidden-false.html:
* LayoutTests/accessibility/ios-simulator/table-ancestory-expected.txt: Removed.
* LayoutTests/accessibility/ios-simulator/table-ancestory.html: Removed.
* LayoutTests/accessibility/mac/aria-grid-with-strange-hierarchy.html:
* LayoutTests/accessibility/mac/aria-hidden-changes-for-non-ignored-elements-expected.txt:
* LayoutTests/accessibility/mac/aria-hidden-changes-for-non-ignored-elements.html:
* LayoutTests/accessibility/mac/selected-text-range-unconnected-object.html:
* LayoutTests/accessibility/out-of-bounds-rowspan.html:
Re-write several tests to be more async.

* LayoutTests/accessibility/table-ancestry-expected.txt: Added.
* LayoutTests/accessibility/table-ancestry.html: Added.
This test used to be iOS-only — it has been ported to macOS too with
this patch.

* LayoutTests/platform/glib/TestExpectations: Skip accessibility/table-ancestry.html.
* LayoutTests/platform/ios/TestExpectations: Enable accessibility/table-ancestry.html.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::unignoredChildren):
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::parentObjectUnignored const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::onlyAddsUnignoredChildren const):
(WebCore::Accessibility::enumerateDescendantsIncludingIgnored):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::nodeIsRowGroup):
(WebCore::AXObjectCache::handleTextChanged):
(WebCore::AXObjectCache::onInertOrVisibilityChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::addRelation):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp:
(WebCore::AccessibilityARIAGridRow::parentObjectUnignored const): Deleted.
* Source/WebCore/accessibility/AccessibilityARIAGridRow.h:
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::updateRoleAfterChildrenCreation):
(WebCore::AccessibilityList::determineAccessibilityRole):
(WebCore::AccessibilityList::determineAccessibilityRoleWithCleanChildren):
* Source/WebCore/accessibility/AccessibilityList.h:
* Source/WebCore/accessibility/AccessibilityListBox.cpp:
(WebCore::AccessibilityListBox::setSelectedChildren):
(WebCore::AccessibilityListBox::selectedChildren):
(WebCore::AccessibilityListBox::visibleChildren):
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
(WebCore::AccessibilityMenuListPopup::selectedChildren):
(WebCore::AccessibilityMenuListPopup::didUpdateActiveOption):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::parentObject const):
(WebCore::AccessibilityNodeObject::boundingBoxRect const):
(WebCore::AccessibilityNodeObject::addChildren):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::dependsOnTextUnderElement const):
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::documentLinks):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::addChildren):
(WebCore::AccessibilityTable::computeCellSlots):
* Source/WebCore/accessibility/AccessibilityTable.h:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::parentObjectUnignored const): Deleted.
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableColumn.cpp:
(WebCore::AccessibilityTableColumn::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp:
(WebCore::AccessibilityTableHeaderContainer::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::objectChangedIgnoredState):
(WebCore::AXIsolatedTree::updatePropertiesForSelfAndDescendants):
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::AXIsolatedTree::updateChildren):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::isUnconnectedNode const):
(WebCore::AXIsolatedTree::objectBecameIgnored):
(WebCore::AXIsolatedTree::objectBecameUnignored):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(attributeValueForTesting):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::attributedStringForElement):
(WTR::AccessibilityUIElement::isInTable const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::attributeValue):
(WTR::AccessibilityUIElement::isInTable const):

Canonical link: <a href="https://commits.webkit.org/286511@main">https://commits.webkit.org/286511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fee40198d7dc25ce7f05967190bef812f52c8744

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28733 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80343 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27102 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77952 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3159 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59471 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17621 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78903 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49333 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65133 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39831 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46732 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22616 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25429 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67849 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81797 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3205 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2009 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67692 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66998 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16771 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10940 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9069 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3155 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5961 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3176 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3183 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1661,9 +1661,13 @@ AXCoreObject::AccessibilityChildrenVector AccessibilityRenderObject::documentLin
     AccessibilityChildrenVector result;
     Document& document = m_renderer->document();
     Ref<HTMLCollection> links = document.links();
+    WeakPtr cache = document.existingAXObjectCache();
```
should we return early here instead of check !cache through each item in loop

twilco replied:
Yes! Will fix that.

fleizach commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1589,6 +1588,18 @@ void enumerateAncestors(const T& object, bool includeSelf, const F& lambda)
+        lambda(object);
+
+    for (const auto& child : object.childrenIncludingIgnored()) {
+        if (child)
```
what allows a child to be returned from this array that is nil?

twilco replied:
It really never should be null...I'll remove the check and see if any tests start to fail (highly doubt it).

We wouldn't have to even consider this check if `AccessibilityChildrenVector` was `Vector<Ref<AXCoreObject>>` (which cannot be null) instead of the current `Vector<RefPtr<AXCoreObject>>`. `RefPtr` doesn't make sense — we either have an object (represented by a `Ref`), or we don't (it's not an entry in the vector).

I've been meaning to make this refactor forever, but it's a big change. Hopefully I can find the time soon.

twilco committed:
AX: Make all layout tests pass on macOS and iOS with ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE)
https://bugs.webkit.org/show_bug.cgi?id=282647
rdar://139315447

Reviewed by Chris Fleizach.

This patch features a variety of changes to make all layout tests pass with ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE):

  1. Prevent tables from returning "invalid" children — non-rows, columns, captions, or table header containers.
     We used to get this for free by way of AccessibilityTable::addChildren() only adding these types of children,
     but we cannot do that anymore, since the point of this effort is to add all children, including those that are
     ignored. Computing cell slots is now split out into a separate function: AccessibilityTable::computeCellSlots().

  2. When `inert` or `visibility` changes, re-compute AXPropertyName::IsIgnored for all descendants.

  3. Fix a bug where we don't auto-focus into modals if the author provided a focus element and immediately made
     that focus element un-focusable (e.g. via display:none). This was caught by a layout test, and was only exposed
     by this change by chance due to a slight timing difference.

  4. When an object becomes unignored, recompute all its properties, since we only cache a minimal set of properties for
     ignored objects.

  5. Split AccessibilityList::determineAccessibilityRole into determineAccessibilityRole and determineAccessibilityRoleWithCleanChildren
     The old implementation relied on children being up-to-date, without anything actually upholding that invariant. Now
     this is more explicit, and should behave more predictably.

  6. Fix a few ATSPI-platform guards added in https://bugs.webkit.org/show_bug.cgi?id=281749 that accidentally applied
     to iOS too.

  7. Drive-by fix to remove a completely unnecessary override for AccessibilityTableCell::parentObjectUnignored().

  8. Add new testing-only macOS attribute handling for `_AXIsInTable` in support of new test table-ancestry.html.

* LayoutTests/accessibility/aria-hidden-subtree-expected.txt:
* LayoutTests/accessibility/aria-hidden-subtree.html:
* LayoutTests/accessibility/aria-hidden-updates-alldescendants-expected.txt:
* LayoutTests/accessibility/aria-hidden-updates-alldescendants.html:
* LayoutTests/accessibility/aria-hidden-with-elements.html:
* LayoutTests/accessibility/aria-labelledby-on-password-input.html:
* LayoutTests/accessibility/aria-modal-in-aria-hidden-expected.txt:
* LayoutTests/accessibility/aria-modal-in-aria-hidden.html:
* LayoutTests/accessibility/deleting-iframe-destroys-axcache.html:
* LayoutTests/accessibility/dynamic-aria-hidden-cell-expected.txt:
* LayoutTests/accessibility/dynamic-aria-hidden-cell.html:
* LayoutTests/accessibility/ignore-modals-without-any-content.html:
* LayoutTests/accessibility/input-type-hidden-in-aria-hidden-false-expected.txt:
* LayoutTests/accessibility/input-type-hidden-in-aria-hidden-false.html:
* LayoutTests/accessibility/ios-simulator/table-ancestory-expected.txt: Removed.
* LayoutTests/accessibility/ios-simulator/table-ancestory.html: Removed.
* LayoutTests/accessibility/mac/aria-grid-with-strange-hierarchy.html:
* LayoutTests/accessibility/mac/aria-hidden-changes-for-non-ignored-elements-expected.txt:
* LayoutTests/accessibility/mac/aria-hidden-changes-for-non-ignored-elements.html:
* LayoutTests/accessibility/mac/selected-text-range-unconnected-object.html:
* LayoutTests/accessibility/out-of-bounds-rowspan.html:
Re-write several tests to be more async.

* LayoutTests/accessibility/table-ancestry-expected.txt: Added.
* LayoutTests/accessibility/table-ancestry.html: Added.
This test used to be iOS-only — it has been ported to macOS too with
this patch.

* LayoutTests/platform/glib/TestExpectations: Skip accessibility/table-ancestry.html.
* LayoutTests/platform/ios/TestExpectations: Enable accessibility/table-ancestry.html.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::unignoredChildren):
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::parentObjectUnignored const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::onlyAddsUnignoredChildren const):
(WebCore::Accessibility::enumerateDescendantsIncludingIgnored):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::nodeIsRowGroup):
(WebCore::AXObjectCache::handleTextChanged):
(WebCore::AXObjectCache::onInertOrVisibilityChange):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::addRelation):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp:
(WebCore::AccessibilityARIAGridRow::parentObjectUnignored const): Deleted.
* Source/WebCore/accessibility/AccessibilityARIAGridRow.h:
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::updateRoleAfterChildrenCreation):
(WebCore::AccessibilityList::determineAccessibilityRole):
(WebCore::AccessibilityList::determineAccessibilityRoleWithCleanChildren):
* Source/WebCore/accessibility/AccessibilityList.h:
* Source/WebCore/accessibility/AccessibilityListBox.cpp:
(WebCore::AccessibilityListBox::setSelectedChildren):
(WebCore::AccessibilityListBox::selectedChildren):
(WebCore::AccessibilityListBox::visibleChildren):
* Source/WebCore/accessibility/AccessibilityMenuListPopup.cpp:
(WebCore::AccessibilityMenuListPopup::selectedChildren):
(WebCore::AccessibilityMenuListPopup::didUpdateActiveOption):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::parentObject const):
(WebCore::AccessibilityNodeObject::boundingBoxRect const):
(WebCore::AccessibilityNodeObject::addChildren):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::dependsOnTextUnderElement const):
(WebCore::AccessibilityObject::isIgnoredWithoutCache const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::documentLinks):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::addChildren):
(WebCore::AccessibilityTable::computeCellSlots):
* Source/WebCore/accessibility/AccessibilityTable.h:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::parentObjectUnignored const): Deleted.
* Source/WebCore/accessibility/AccessibilityTableCell.h:
* Source/WebCore/accessibility/AccessibilityTableColumn.cpp:
(WebCore::AccessibilityTableColumn::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityTableHeaderContainer.cpp:
(WebCore::AccessibilityTableHeaderContainer::computeIsIgnored const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::objectChangedIgnoredState):
(WebCore::AXIsolatedTree::updatePropertiesForSelfAndDescendants):
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::AXIsolatedTree::updateDependentProperties):
(WebCore::AXIsolatedTree::updateChildren):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::isUnconnectedNode const):
(WebCore::AXIsolatedTree::objectBecameIgnored):
(WebCore::AXIsolatedTree::objectBecameUnignored):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(attributeValueForTesting):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::attributedStringForElement):
(WTR::AccessibilityUIElement::isInTable const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::attributeValue):
(WTR::AccessibilityUIElement::isInTable const):

Canonical link: https://commits.webkit.org/286511@main
